### DISCUSSION TITLE
Ccache support for windows Visual studio 2019?

### DISCUSSION THREAD

**Discussion by thiyagsdevops** [ccache/ccache#978] [Questions & answers] [answered]
We are planning to bring in ccache for our entire organization and during our POC stage on windows, 

BUILD SYSTEM
Git lab
CODE
C++ and 
COMPILER
C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.28.29333/bin/Hostx64/x64/cl.exe
TOOLS
cmake with Visual studio 2019

With below lines in CMakeLists.txt we were able to prefix ccache with clang compiler on mac with Xcode.
**set(CMAKE_XCODE_ATTRIBUTE_CC "$(SRCROOT)/tools/ccache-clang")
set(CMAKE_XCODE_ATTRIBUTE_CXX "$(SRCROOT)/tools/ccache-clang++")

ccache-clang++:
#!/bin/sh
if type -p ccache >/dev/null 2>&1; then
#export CCACHE_MAXSIZE=100G
export CCACHE_CPP2=true
export CCACHE_HARDLINK=true
export CCACHE_DEPEND=true
export CCACHE_IGNOREOPTIONS="-DDUMMY_VERSION=*"
export CCACHE_SLOPPINESS=file_macro,time_macros,include_file_mtime,include_file_ctime,file_stat_matches
exec ccache clang++ "$@"
else
exec clang++ "$@"
fi

Is there is a similar option available for windows with Visual Studio 2019 as well. Please suggest

**Comment by jrosdahl** [comment, marked as answer]
What ccache supports is documented on [Supported platforms and compilers](https://ccache.dev/platform-compiler-language-support.html). Under _Run-time support_/_Compilers_ you'll find _MSVC (Microsoft Visual C++)_ which links to #506 which most likely will be included in ccache 4.6.

**Reply by thiyagsdevops** [reply]
@jrosdahl Thanks for the update. We started using ccache for clang & clang++ on mac and we see a huge improvement on build timing with ccache. We started using ccache across our organization. Thanks for the great work

**Comment by ranjithp04** [comment]
@jrosdahl how to integrated ccache with vs2019 MSVC on windows. I've tried below steps and it didn't help.

```
PS C:\Users\builder\Downloads\my_project> cat .\CMakeLists.txt
cmake_minimum_required(VERSION 3.22.1)
find_program(CCACHE_EXECUTABLE ccache)
mark_as_advanced(CCACHE_EXECUTABLE)
if(CCACHE_EXECUTABLE)
  foreach(LANG C CXX)
    if(NOT DEFINED CMAKE_${LANG}_COMPILER_LAUNCHER AND NOT CMAKE_${LANG}_COMPILER MATCHES ".*/ccache")
      message(STATUS "Enabling ccache for ${LANG}")
      set(CMAKE_${LANG}_COMPILER_LAUNCHER ${CCACHE_EXECUTABLE} CACHE STRING "")
    endif()
  endforeach()
endif()
project (hello)
add_executable(hello helloworld.cpp)

PS C:\Users\builder\Downloads\my_project\build> cmake ..
-- Building for: Visual Studio 16 2019
-- Enabling ccache for C
-- Enabling ccache for CXX
-- The C compiler identification is MSVC 19.28.29336.0
-- The CXX compiler identification is MSVC 19.28.29336.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.28.29333/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.28.29333/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: C:/Users/builder/Downloads/my_project/build

PS C:\Users\builder\Downloads\my_project\build> cmake --build .
Microsoft (R) Build Engine version 16.8.3+39993bd9d for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

  Checking Build System
  Building Custom Rule C:/Users/builder/Downloads/my_project/CMakeLists.txt
  helloworld.cpp
  hello.vcxproj -> C:\Users\builder\Downloads\my_project\build\Debug\hello.exe
  Building Custom Rule C:/Users/builder/Downloads/my_project/CMakeLists.txt

PS C:\Users\builder\Downloads\my_project\build> ccache.exe -vs
Summary:
  Cache directory:  C:\Users\builder\AppData\Roaming/ccache
  Primary config:   C:\Users\builder\AppData\Roaming/ccache/ccache.conf
  Secondary config: /usr/local/etc/ccache.conf
  Stats updated:    02/28/22 13:18:28
  Hits:                0 /      0
    Direct:            0 /      0
    Preprocessed:      0 /      0
  Misses:              0
    Direct:            0
    Preprocessed:      0
Primary storage:
  Hits:                0 /      0
  Misses:              0
  Cache size (GB):  0.00 / 100.00 (0.00 %)
  Files:               0
PS C:\Users\builder\Downloads\my_project\build>
```

**Reply by cristianadam** [reply]
That's because CMake doesn't support `CMAKE_<LANG>_COMPILER_LAUNCHER` for Visual Studio projects. See the documentation https://cmake.org/cmake/help/latest/prop_tgt/LANG_COMPILER_LAUNCHER.html

On Windows with MSVC it only works with "NMake Makefiles", "NMake Makefiles JOM", "Ninja" or "Ninja Multi-Config" generators.

**Reply by cristianadam** [reply]
There is an upstream CMake issue at https://gitlab.kitware.com/cmake/cmake/-/issues/19326 and apparently one could use `ccache` in the same way `clang-cl` is being used for MSBuild instead of cl.

**Reply by saudet** [reply]
Yes, that works just fine for me. I had already found that setting the CC and CXX environment variables to "cl.exe" works best with Ninja for any projects using CMake, so I gave it a try simply by setting them to "ccache.exe cl.exe" instead and it actually works! See example here:
https://github.com/bytedeco/javacpp-presets/blob/master/pytorch/cppbuild.sh#L103
